tiledb/api/c_api/vfs/vfs_api.cc

```diff
@@ -297,6 +297,30 @@ capi_return_t tiledb_vfs_ls(
   return TILEDB_OK;
 }
 
+capi_return_t tiledb_vfs_ls_recursive(
+    tiledb_vfs_t* vfs,
+    const char* path,
+    int32_t (*callback)(const char*, void*, void*),
+    void* data,
+    void* data_offsets,
+    int64_t max_paths) {
+  ensure_output_pointer_is_valid(data);
+  ensure_output_pointer_is_valid(data_offsets);
+  auto children = vfs->ls_recursive(tiledb::sm::URI(path), max_paths);
+
+  // Apply first callback with empty path to set first offset of 0.
+  int rc = callback("", data, data_offsets);
```

Comment:
This should go. You're coordinating your callback implementation with the callback API. If the callback needs to know its first invocation it should track that data itself or the callback API should be changed dramatically to have a type of invocation (though I don't think that is needed here).